src/evict.c

```diff
@@ -550,21 +575,28 @@ int performEvictions(void) {
             struct evictionPoolEntry *pool = EvictionPoolLRU;
 
             while(bestkey == NULL) {
-                unsigned long total_keys = 0, keys;
 
                 /* We don't want to make local-db choices when expiring keys,
                  * so to start populate the eviction pool sampling keys from
                  * every DB. */
-                for (i = 0; i < server.dbnum; i++) {
-                    db = server.db+i;
-                    dict = (server.maxmemory_policy & MAXMEMORY_FLAG_ALLKEYS) ?
-                            db->dict : db->expires;
-                    if ((keys = dictSize(dict)) != 0) {
-                        evictionPoolPopulate(i, dict, db->dict, pool);
-                        total_keys += keys;
-                    }
+                int samples = 0;
+                struct EvictDB* dbs = (struct EvictDB*)zmalloc(sizeof(struct EvictDB)*server.dbnum);
+                int count = getDbsToEvict(dbs);
+                int i;
+                if (!count) {
+                    zfree(dbs);
+                    break; /* No keys to evict. */
+                }
+                for (i=0; i < count; i++) {
+                    samples += evictionPoolPopulate(
+                            dbs[i].dbid,
+                            dbs[i].sampleDict,
+                            dbs[i].keyDict,
+                            pool,
+                            (server.maxmemory_samples+count-1)/count);
+                    if (samples >= server.maxmemory_samples) break;
```

Comment:
@madolson I don't think this change is necessary.
The main reason why this modification can improve performance is here, it will break when maxmemory_samples is reached, but it will cause evict to be very unfair. https://github.com/redis/redis/commit/e423f76e75bea83106c271b56d516c20cf74a16a This change is also to ensure fairness between multiple dbs.
If we really want to improve the speed of eviction, we can reduce the value of  ```maxmemory-samples``` and ```maxmemory-eviction-tenacity```.